File: src/main/java/store/sales/domain/model/discounts/DiscountBuilder.java
package store.sales.domain.model.discounts;

import java.time.LocalDateTime;

public class DiscountBuilder {
    private long id;
    private LocalDateTime expirationTime;
    private DiscountStrategy strategy;
    private int maxDiscountValue;

    public DiscountBuilder setId(long id) {
        this.id = id;
        return this;
    }

    public DiscountBuilder setExpirationTime(LocalDateTime expirationTime) {
        this.expirationTime = expirationTime;
        return this;
    }

    public DiscountBuilder setStrategy(DiscountStrategy strategy) {
        this.strategy = strategy;
        return this;
    }

    public DiscountBuilder setMaxDiscountValue(int maxDiscountValue) {
        this.maxDiscountValue = maxDiscountValue;
        return this;
    }

    public Discount build() {
        return new Discount(id, expirationTime, strategy, maxDiscountValue);
    }
}

File: Store/test/Sales.Domain.Tests/TestHelpers/DiscountBuilder.cs
using System;
using Sales.Domain.Model.Discounts;

namespace Sales.Domain.Tests.TestHelpers
{
    public class DiscountBuilder
    {
        private long _id;
        private DateTime _expirationDate;
        private DiscountStrategy _discountStrategy;
        private long? maxDiscountValue;
        public DiscountBuilder()
        {
            _id = new Random().Next(1, int.MaxValue);
            _expirationDate = DateTime.Now.AddYears(1);
            _discountStrategy = new ValueBasedDiscount(0);
            maxDiscountValue = null;
        }

        public DiscountBuilder AsValueBasedDiscount(long value)
        {
            _discountStrategy = new ValueBasedDiscount(value);
            return this;
        }

        public DiscountBuilder AsPercentageBasedDiscount(float percent)
        {
            _discountStrategy = new PercentageBasedDiscount(percent);
            return this;
        }

        public DiscountBuilder WithMaxDiscountValue(long value)
        {
            maxDiscountValue = value;
            return this;
        }

        public DiscountBuilder WithExpireDateAs(DateTime date)
        {
            this._expirationDate = date;
            return this;
        }

        public Discount Build()
        {
            return new Discount(_id, _expirationDate, _discountStrategy, maxDiscountValue);
        }
    }
}